src/os_posix/os_fs.c

```diff
@@ -789,3 +971,176 @@ __wt_os_posix(WT_SESSION_IMPL *session)
 
     return (0);
 }
+
+/*
+ * Here is the synchronization protocol to prevent race conditions when a session is remapping the
+ * file while others might be reading or writing it:
+ *
+ * Every time someone reads or writes from the mapped region, they increment the "use" count via
+ * cas. If someone wants to change the file size, they set the "stop" flag. If a session sees
+ * the stop flag, it does not read via mmap, but resorts to the regular syscall. The session
+ * that set the stop flag spin-waits until the "use" count goes to zero. Then it changes the
+ * file size and remaps the region without synchronization. Once all that is done, it resets the
+ * "stop" flag.
+ */
+/*
+ * __wt_drain_mmap_users --
+ *     Wait until all the sessions using the memory mapped region are done.
+ */
+void
+__wt_drain_mmap_users(WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session)
+{
+
+    WT_FILE_HANDLE_POSIX *pfh;
+    WT_SESSION_IMPL *session;
+    bool cas_result;
+    uint32_t old_state;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+    pfh = (WT_FILE_HANDLE_POSIX *)file_handle;
+
+    __wt_verbose(session, WT_VERB_FILEOPS, "%s, drain-mmap-users: buffer=%p\n", file_handle->name,
+      (void *)pfh->mmap_buf);
+
+    WT_ASSERT(session, S2C(session)->mmap);
+    WT_ASSERT(session, pfh->mmap_file_mappable);
+
+wait:
+    /* Wait until it looks like no one is resizing the region */
+    while ((old_state = pfh->mmap_resizing) == 1)
+        WT_PAUSE();
+
+    if ((cas_result = __wt_atomic_casv32(&pfh->mmap_resizing, old_state, 1)) == false)
+        goto wait;
+
+    /*
+     * Wait for any sessions using the region for I/O to finish. Now that we have set the resizing
+     * flag, new sessions will not use the region, defaulting to system calls instead.
+     */
+    while (pfh->mmap_usecount > 0)
+        WT_PAUSE();
+}
```

Comment:
Static analyzers are going to ding you for `cas_result` being set but never read.

I'd suggest:
```
while (pfh->mmap_resizing != 0 || !__wt_atomic_casv32(&pfh->mmap_resizing, 0, 1)) 
    WT_PAUSE();
```

Reply:
On second thought, this should probably be `__wt_yield()`, not `WT_PAUSE`, we want other processes to get the CPU.

Reply:
I think this should be okay. cas_result is compared to "false" after it is set (at the very end of the if-condition.

Not sure about using __wt_yield()... This is meant to be a very short critical section, so whoever has set the counter is likely to be on the CPU. I think we'll get better mileage out of WT_PAUSE.

Regarding your while-loop suggestion, I'd prefer not to "cas" in a while loop, because this will cause very bad invalidation traffic among caches. At the moment, I have it implemented in a "test and test-and-set" fashion, where we spin on the old_state variable until it "looks" free and then try to cas it.

Reply:
> I think this should be okay. cas_result is compared to "false" after it is set (at the very end of the if-condition.

Agreed, I was wrong.

> Not sure about using __wt_yield()... This is meant to be a very short critical section, so whoever has set the counter is likely to be on the CPU. I think we'll get better mileage out of WT_PAUSE.

I don't follow your point, can you please explain further? In other words, why do we think the set of the counter happened on this CPU? It could be any CPU, why is the current CPU more likely?

The problem with `WT_PAUSE` is we're burning CPU to no purpose, at some point. If you think this needs to be optimized, I'd suggest `__wt_spin_backoff()`, which spins, then yields, then sleeps.

> Regarding your while-loop suggestion, I'd prefer not to "cas" in a while loop, because this will cause very bad invalidation traffic among caches. At the moment, I have it implemented in a "test and test-and-set" fashion, where we spin on the old_state variable until it "looks" free and then try to cas it.

I'm suggestion a test and test-and-set fashion, at least I think I am. Aren't we testing first?

Reply:
> I don't follow your point, can you please explain further? In other words, why do we think the set of the counter happened on this CPU? It could be any CPU, why is the current CPU more likely?
>

Sorry, what I meant was that the counter could be set on any CPU, but whoever set it will likely unset it very soon (before they are descheduled), so it makes more sense for us to pause on the CPU as opposed to yielding, which is a very expensive operation. 

> The problem with `WT_PAUSE` is we're burning CPU to no purpose, at some point. If you think this needs to be optimized, I'd suggest `__wt_spin_backoff()`, which spins, then yields, then sleeps.
> 

We are certainly burning the CPU, but we are expecting that we won't be burning this for long, because the amount of work that happens in the other thread between setting the flag and unsetting it is very small. 

I agree that __wt_spin_backoff() might be a good option. If we were right about the other thread resetting the counter soon, then we will proceed after spinning for a bit, but if we were wrong we will yield the CPU. I'd like to test that option to see if it has any effect on performance. 

> > Regarding your while-loop suggestion, I'd prefer not to "cas" in a while loop, because this will cause very bad invalidation traffic among caches. At the moment, I have it implemented in a "test and test-and-set" fashion, where we spin on the old_state variable until it "looks" free and then try to cas it.
> 
> I'm suggestion a test and test-and-set fashion, at least I think I am. Aren't we testing first?


`while (pfh->mmap_resizing != 0 || !__wt_atomic_casv32(&pfh->mmap_resizing, 0, 1))`

Ok yes, you are right. Your suggested loop *is* test-and-test-and-set. I didn't see this first. This is more difficult to read for me than the exiting code, and I am paranoid about getting lock-free synchronization right. :)  Ultimately, this is just my personal opinion, so feel free to implement your change!

Reply:
If the thread we're waiting for is on our CPU, then we're going to sit and spin until the scheduler gives up?

I probably wouldn't fight hard about the existing loop, but it's significantly more complicated plus it requires a `goto`...

Reply:
> If the thread we're waiting for is on our CPU, then we're going to sit and spin until the scheduler gives up?
> 
With `WT_PAUSE` we certainly would. With `__wt_spin_backoff() `-- only for a bit. I'll test if there are any performance differences with `__wt_spin_backoff()`  vs. `WT_PAUSE`. I don't expect any.

> I probably wouldn't fight hard about the existing loop, but it's significantly more complicated plus it requires a `goto`...

I'll take another look.

Reply:
I ran the experiments. There's no performance difference overall if we replace WT_PAUSE with __wt_spin_backoff(). So I committed the code with __wt_spin_backoff(). I also cleaned up the loop, though I kept the goto statement. I think this helps readability. Please take another look.